### DISCUSSION TITLE
Regression since v0.44.1 - Error 422 if `Form` is declared before `File`

### DISCUSSION THREAD

**Discussion by thomasleveil** [fastapi/fastapi#9116] [Questions] [answered]
Here's a [branch](https://github.com/thomasleveil/fastapi/tree/file_and_form_together) with my use case:

```python
from starlette.testclient import TestClient

from fastapi import FastAPI, File, Form

app = FastAPI()


@app.post("/file_before_form")
def file_before_form(
    file: bytes = File(...), city: str = Form(...),
):
    return {"file_content": file, "city": city}


@app.post("/file_after_form")
def file_after_form(
    city: str = Form(...), file: bytes = File(...),
):
    return {"file_content": file, "city": city}


client = TestClient(app)


def test_file_before_form():
    response = client.post(
        "/file_before_form", data={"city": "Thimphou"}, files={"file": "<file content>"}
    )
    assert response.status_code == 200, response.text
    assert response.json() == {"file_content": "<file content>", "city": "Thimphou"}


def test_file_after_form():
    response = client.post(
        "/file_after_form", data={"city": "Thimphou"}, files={"file": "<file content>"}
    )
    assert response.status_code == 200, response.text  # <-------------- this fails since ab2b86f
    assert response.json() == {"file_content": "<file content>", "city": "Thimphou"}

```

### Description

Upgrading FastApi from 0.42.0 to 0.61.0 on one of my project, I stumbled upon a regression in which FastApi will respond with HTTP Error 422 when ever posting a form-data request to an endpoint where a `File` is declared **after** a `Form`.

Searching the Internet, I found another user suffering from this issue : https://stackoverflow.com/questions/61376754/fastapi-files-must-be-loaded-before-form-in-function-parameters

I've created a branch with tests to showcase the issue : https://github.com/thomasleveil/fastapi/tree/file_and_form_together

### Environment

* OS: Linux
* FastAPI Version 0.44.1 and later
* Python version: 3.8.5

### Additional context

 I've bisected the repo and figured out the regression was introduced in commit ab2b86fe2ce8fe15e91aaec179438e24ff7b7ed0

**Comment by HacKanCuBa** [comment]
*[EDIT: dismiss this comment]*

After almost 2 hours of banging my head against the desktop, I end up concluding that the problem I was facing had to be a FastAPI bug. So I came here and found this issue and voilá! It was a bug.

But that's not all! Because my problem wasn't __exactly__ the same. Take a look at this sample code:

```python
router = APIRouter()

router.post('/{id}/')
async def update_blah(request: Request, id: int, something: str = Form(...)):
   ...

router.post('/other/')
async def other_blah(request: Request, file: UploadFile = File(...)):
   ...
```

And **that** was failing when POSTing to `/other/` with 422 reporting errors as if it were posted to the `update_blah` endpoint. The solution? As @thomasleveil said: file before form.

```python
router = APIRouter()

router.post('/other/')
async def other_blah(request: Request, file: UploadFile = File(...)):
   ...

router.post('/{id}/')
async def update_blah(request: Request, id: int, something: str = Form(...)):
   ...
```

Can't believe it? I couldn't either. This is some really crazy bug, I'm gonna take a look into it later today :)

FastAPI version: `0.61.1`

**Comment by thomasleveil** [comment]
@HacKanCuBa what you describe is not a bug and the behavior is clearly described in the documentation at https://fastapi.tiangolo.com/tutorial/path-params/#order-matters

**Comment by HacKanCuBa** [comment]
Damn, you are totally right!!! it's because `/other/` *could* be an `id` for the other route! I need to sleep more xD
Thanks for the quick reply!

**Comment by perkfly** [comment]
This is still not fixed as of May 2021(0.65.1), using `UploadFile` as the type seems to be ok, though. Any updates on this issue?

**Comment by AKlaus** [comment]
Confirm that the issue still exists in `FastAPI v0.70.1` (used `Python v3.7.10`).

Tested by running cURL (to avoid accusations towards Swagger):
```
curl --location --request POST 'http://127.0.0.1:5001/file_after_form' --form 'city="fff"' --form 'file=@"file.png"'
```
that gets _HTTP code 422_ with message:
```
{"detail":[{"loc":["body","file"],"msg":"byte type expected","type":"type_error.bytes"}]}
```

(I raised #4384 initially, but closing it as a duplicate).

**Comment by mht2953658596** [comment]
Did anyone get the reason？puzzled

**Comment by thomasleveil** [comment]
I rebased my branch onto 0.110.0 (e40747f10ae911910e9cb9a9684576f3b21304c9) and this behavior can still be observed in FastAPI v1.110.0

See test case https://github.com/thomasleveil/fastapi/blob/5af0a0908a638af698a3b5d1464402d3b9a34b55/tests/test_file_and_form.py

Tested with : 
- Python 3.12.1
- FastAPI v0.110.0
- Pydantic 2.6.2

**Comment by YuriiMotov** [comment]
There is a PR #9410 that fixes this issue, but it has to be modified a bit to work on current FastAPI version.
If you are interested in fixing this issue, please review or just upvote that PR

**Reply by thomasleveil** [reply]
Thank you for pointing this PR to me 🙏
I'm trying this PR right now after rebasing onto 1.110.0, but it fails. Also I added a test case with multiple files upload and the PR is not fixing that specific case. 
I'll spend a few more hours on this and see if I can come up with a PR

**Reply by YuriiMotov** [reply]
Have you made changes I proposed in my review? I just run this on 0.110.0 and it works (with the changes I mentioned in the review). I also tried creating the endpoint with 2 files and it also works

**Reply by thomasleveil** [reply]
indeed removing the assertion the the exact error message fixes those. However I have another test which still fails, but that might be a different issue (it involves Json type). 
Anyway, I'm polishing up my branch and will submit a PR

**Reply by YuriiMotov** [reply]
> I have another test which still fails, but that might be a different issue (it involves Json type)

Can you show a testcase that fails?

**Reply by thomasleveil** [reply, marked as answer]
I submitted  #11194 and will open a new discussion for the case with `RootModel[Json[ImagesProperty]]` which still is affected by the order in which the parameters are declared
